src/Messages.jl

```diff
@@ -197,29 +199,30 @@ Represents a HTTP Request Message.
 - `headers::Vector{Pair{String,String}}`
    [RFC7230 3.2](https://tools.ietf.org/html/rfc7230#section-3.2)
 
-- `body::Vector{UInt8}`
+- `body::Union{Vector{UInt8}, IO}`
    [RFC7230 3.3](https://tools.ietf.org/html/rfc7230#section-3.3)
 
 - `response`, the `Response` to this `Request`
 
-- `txcount`, number of times this `Request` has been sent (see RetryRequest.jl).
-
 - `parent`, the `Response` (if any) that led to this request
   (e.g. in the case of a redirect).
    [RFC7230 6.4](https://tools.ietf.org/html/rfc7231#section-6.4)
 
+- `context`, a `Dict{Symbol, Any}` store used by middleware to share state
+
 You can get each data with [`HTTP.method`](@ref), [`HTTP.headers`](@ref), [`HTTP.uri`](@ref), and [`HTTP.body`](@ref).
 
 """
-mutable struct Request <: Message
+mutable struct Request{T} <: Message
     method::String
     target::String
     version::VersionNumber
     headers::Headers
-    body::Vector{UInt8}
+    body::T # Vector{UInt8} or some kind of IO
     response::Response
-    txcount::Int
-    parent
+    url::URI
```

Comment:
We add a new `url` field because there are many times throughout the client layers where we want access to the original url provided by the user and not just the `resource(url)` that we store as the `target` field.